mantidproject/mantid
# Reenable gitHub api for nightly builds

Add read only authentication to github api calls, reenable update checking and instrument downloading for nightly builds (This makes sense where people take a nightly build but then stay on that build for months).

With that later commit this will also fail over to using the anonymous calls when the token fails or runs out of calls.

**To test:**
If you have a non installer build of Mantid you will need to enable version checking and instrument downloading in the mantid.properties file, if you have an installer based version it will already be working.

```
CheckMantidVersion.OnStartup = 1
UpdateInstrumentDefinitions.OnStartup = 1
```

This command lets you check the current rate limit usage (**make sure you use the right token for your environment**).

``` bash
curl -H "Authorization: token 8ec7afc857540ee60af78cba1cf7779a6ed0b6b9" https://api.github.com/rate_limit
```

**The possible tokens (that one is windows) are visible in the cmakelist file of this PR**.
1. Startup mantid
2. Set the log level to debug
3. close and restart mantid
4. You should catch debug log lines like

```
GitHub API 4999 of 5000 calls left. Resets at 1 June 2016 10:24:34 GMT
```
1. you can run the rate check line above to verify the correct token is being used.
2. You can also force a github api call using `DownloadInstrument(ForceRefresh=True)`
### This next bit needs to be complete within 1 hour (or the rate limit refreshes)
1. Next exhaust the github api limit by running this in a gitbash window (or normal bash window), it took about 45 min for me, you can track progress using the rate limit call in another window **(again make sure you are using the right token).**

``` bash
for i in {1..5000}; do curl -H "Authorization: token 8ec7afc857540ee60af78cba1cf7779a6ed0b6b9" https://api.github.com/repos/mantidproject/mantid/releases/latest; done
```

To exhaust the anonymous calls

``` bash
for i in {1..60}; do curl https://api.github.com/repos/mantidproject/mantid/releases/latest; done
```
1. Delete github.json from the instrument cache directory (windows %appdata%\mantidproject\mantid\instrument, Linux .mantid\instrument)
2. Then startup mantid again.
3. You should get a warning like this, but it should start fine.

```
InternetError:  [304] The Github API rate limit has been reached, try again after 1 June 2016 10:24:34 GMT
```

Fixes #16519.

Does not need to be in the release notes as it only affects nightly builds (in a way that is noticeable).

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
